### PR TITLE
feat: relations and expressions binary output for the textplan parser

### PR DESCRIPTION
**Pull request by EpsilonPrime**
* Relations are now interconnected as defined by pipelines.
    * Expressions are now emitted to binary plans.
    * HasErrors matcher now explains the difference in errors.

### CODE REVIEW

File: src/substrait/textplan/parser/tests/TextPlanParserTest.cpp [hunk 1]
```diff
@@ -117,7 +117,15 @@ std::vector<TestCase> getTestCases() {
               WhenSerialized(EqSquashingWhitespace(R"(pipelines {
                 myself -> a;
                 b -> myself;
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/tests/TextPlanParserTest.cpp:119]: Filed issue #64.

File: src/substrait/textplan/parser/tests/TextPlanParserTest.cpp [hunk 2]
```diff
@@ -665,26 +689,160 @@ std::vector<TestCase> getTestCases() {
...
+          AsBinaryPlan(Partially(EqualsProto<::substrait::proto::Plan>(
+              R"(relations { root { input { project {
+                expressions { cast { type { i32 {} } input { literal { i8: 123 } } } }
+                expressions { cast { type { i64 {} } input { cast { type { i32 {} } input { literal { i8: 123 } } } } } }
+              } } } })"))),
+      },
+      {
+          "test13-functions",
+          R"(extension_space blah.yaml {
+            function sum:str as sum;
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/tests/TextPlanParserTest.cpp:713]: Updated.

File: src/substrait/textplan/parser/tests/TextPlanParserTest.cpp [hunk 3]
```diff
@@ -665,26 +689,160 @@ std::vector<TestCase> getTestCases() {
...
+            base_schema schema;
+          }
+
+          read relation read2 {
+            source named2;
+            base_schema schema2;
+          }
+
+          join relation join {
+            //type JOIN_TYPE_UNSPECIFIED;
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/tests/TextPlanParserTest.cpp:752]: Was waiting for support, I've replaced this with the future syntax which is currently just getting swallowed.

File: src/substrait/textplan/parser/SubstraitPlanRelationVisitor.cpp [hunk 4]
```diff
@@ -378,10 +379,46 @@ std::any SubstraitPlanRelationVisitor::visitRelationExpression(
   return defaultResult();
 }
 
+std::any SubstraitPlanRelationVisitor::visitExpression(
+    SubstraitPlanParser::ExpressionContext* ctx) {
+  if (dynamic_cast<SubstraitPlanParser::ExpressionFunctionUseContext*>(ctx)) {
+    return visitExpressionFunctionUse(
+        dynamic_cast<SubstraitPlanParser::ExpressionFunctionUseContext*>(ctx));
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/SubstraitPlanRelationVisitor.cpp]: Done.

File: src/substrait/textplan/SymbolTablePrinter.cpp [hunk 5]
```diff
@@ -432,6 +435,199 @@ std::string SymbolTablePrinter::outputToText(const SymbolTable& symbolTable) {
...
+    const SymbolInfo& symbolInfo,
+    ::substrait::proto::Rel* relation) {
+  auto relationData = ANY_CAST(std::shared_ptr<RelationData>, symbolInfo.blob);
+
+  // Connect up the incoming inputs in the relation data to the appropriate
+  // input/left/right/inputs of this relation (which recursively also needs to
+  // have its inputs added).
+  if (symbolInfo.subtype.type() != typeid(RelationType)) {
+    // The type isn't one we expected bail.
+    return;
```
Comment (westonpace) [inline src/substrait/textplan/SymbolTablePrinter.cpp:448]: We might consider some kind of `UNREACHABLE` macro which does nothing in release mode but throws an exception (or asserts with the line number) in debug mode.

File: src/substrait/textplan/SymbolTablePrinter.cpp [hunk 6]
```diff
@@ -432,6 +435,199 @@ std::string SymbolTablePrinter::outputToText(const SymbolTable& symbolTable) {
...
+            continuingRelationData->relation;
+        addInputsToRelation(
+            *relationData->continuingPipeline,
+            relation->mutable_project()->mutable_input());
+      }
+      break;
+    case RelationType::kJoin:
+      if (relationData->newPipelines.size() == kBinaryRelationInputCount) {
+        auto leftRelationData = ANY_CAST(
+            std::shared_ptr<RelationData>, relationData->newPipelines[0]->blob);
```
Comment (westonpace) [inline src/substrait/textplan/SymbolTablePrinter.cpp:470]: So this means...

```
read1 -> project1 -> join1
read2 -> project2 -> join1
```

In this case `project1` will be the `left` and `project2` will be the `right`?

**Review by westonpace — APPROVED** [review]

**Pull request merged**